From 0ff455d1d446e34ae6c2596d4d8491f66fc61913 Mon Sep 17 00:00:00 2001
From: Manish Dharanenthiran <quic_mdharane@quicinc.com>
Date: Sat, 2 Dec 2023 03:38:31 +0530
Subject: [PATCH] wifi: mac80211: Fix memory corruption during mesh beacon
 update

During mesh beacon update, u64 flag is used to check for
bit set/unset for validation purpose. But, in
'ieee80211_mbss_info_change_notify' API, unsigned long flag
is modified using sizeof(u64). This leads to following issue:

 > 'mbss_changed' flag in 'ieee80211_if_mesh' is declared as
    unsigned_long which creates an architecture dependency
    (32bit vs 64bit) while modifying it with u64 flag which
    leads to memory corruption.

Fix above mentioned issue by replacing unsigned long with u64
for changed flag.

Signed-off-by: Manish Dharanenthiran <quic_mdharane@quicinc.com>
---
 net/mac80211/ieee80211_i.h | 2 +-
 net/mac80211/mesh.c        | 2 +-
 2 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/net/mac80211/ieee80211_i.h b/net/mac80211/ieee80211_i.h
index e8b76ad..6550e47 100644
--- a/net/mac80211/ieee80211_i.h
+++ b/net/mac80211/ieee80211_i.h
@@ -698,7 +698,7 @@ struct ieee80211_if_mesh {
 	struct timer_list mesh_path_root_timer;
 
 	unsigned long wrkq_flags;
-	unsigned long mbss_changed;
+	u64 mbss_changed;
 
 	bool userspace_handles_dfs;
 
diff --git a/net/mac80211/mesh.c b/net/mac80211/mesh.c
index 6001f0d..3d3e315 100644
--- a/net/mac80211/mesh.c
+++ b/net/mac80211/mesh.c
@@ -1204,7 +1204,7 @@ void ieee80211_mbss_info_change_notify(struct ieee80211_sub_if_data *sdata,
 				       u64 changed)
 {
 	struct ieee80211_if_mesh *ifmsh = &sdata->u.mesh;
-	unsigned long bits = changed;
+	u64 bits = changed;
 	u32 bit;
 
 	if (!bits)
-- 
2.17.1

